keps/sig-auth/3276-raw-request-and-response-bodies-in-audit/README.md

```diff
@@ -0,0 +1,606 @@
+
+## Summary
+
+The current audit logs include serialized and deserialized request and response bodies when using `Request` and
+`RequestResponse` levels. During the serialization process, it is possible to loss some important information.
+Such a functionality will be very helpful when investigating issues similar to
+[CVE-2019-11253](https://github.com/kubernetes/kubernetes/issues/83253).
+
+## Motivation
+
+### Goals
+
+* Include raw request and response bodies in the audit logs in the API Server.
+
+### Non-Goals
+
+* N/A
+
+## Proposal
+
+### User Stories (Optional)
+
+* As a cluster-admin, I need a way to investigate raw requests and responses exchanged with the API Server, so that
+I can debug failures more effectively.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+* Sensitive data may be compromised as the mechanism will not employ any sensitive data masking process.
+
+## Design Details
+
+There are two major pieces of the implementation. The first one is extending the audit `Event` and the others
+is how to enable this feature.
+
+### Extending the audit Event
+
+As for the former, the `Event` struct will be extended by adding two new fields:
+
+```go
+type Event struct {
+  ...
+
+  // Raw body from the request, in base-64-encoded format. The RawRequestObject is recorded as-is in the request.
+	// Only logged at Request Level and higher.
+	// +optional
+	RawRequestObject string
+	// Raw body from the request, in base-64-encoded format. The RawRequestObject is recorded as-is in the response.
+	// Omitted for non-resource requests.  Only logged at Response Level.
+	// +optional
+	RawResponseObject string
+
+  ...
+}
+```
+
+The `RawRequestObject` will be recorded in the `NewEventFromRequest` function, whereas the `RawResponseObject` will
+be recorded in `LogResponseObject` function.
+
+### Turning this feature on
+
+Proposal 1:
+
+We'll introduce a new auditing level, called `RawRequestResponse`. Once it's on, it will turn the raw request and response
+audit logging on and disable using `ResponseObject` and `RequestObject` fields from the audit `Event`. The idea is to make
+both features mutually exclusive.
+
+The above approach will require activation on the Policy level, such as:
+
+```yaml
+# TODO: Investigate if we need to iterate the API in order to introduce a new level
+# Since level is mapped to a string - maybe we don't need to?
+apiVersion: audit.k8s.io/v2alpha1
+kind: Policy
+rules:
+  # Log pod changes at RawRequestResponse level
+  - level: RawRequestResponse
+    resources:
+    - group: ""
+      resources: ["pods"]
+  ...
+```
+
+Proposal 2:
+
+We will introduce a new flag to the API Server instrumenting it to log raw instead of serialized requests and
+responses. The flag will be called `-audit-log-raw-requests-response`
+
+Proposal 3:
+
+We will extend the `PolicyRule` struct and introduce additional field containing a generic configuration attributes:
```

Comment:
I'd recommend something like this approach, but make it a typed fielded rather than a generic string.